[webpack] Production sourcemap for javascript

Hello @greypants, i am unable to generate a sourcemap for javascript in production. I simply put the devtool option directly to webpackConfig and removed the one in the development conditional.

``` js
var webpackConfig = {
   ...,
   devtool: 'source-map',
   ...
}
```

the error in terminal's stream

``` sh
path.js:8
    throw new TypeError('Path must be a string. Received ' +
    ^

TypeError: Path must be a string. Received [ 'main-7cdd5e7952f0ad06bc45.js',
  'main-7cdd5e7952f0ad06bc45.js.map' ]
    at assertPath (path.js:8:11)
    at Object.posix.join (path.js:479:5)
    at Compiler.<anonymous> (/Users/tsm/Sites/__projects/backbone-starter/gulpfile.js/lib/webpackManifest.js:15:66)
    at Compiler.applyPlugins (/Users/tsm/Sites/__projects/backbone-starter/node_modules/tapable/lib/Tapable.js:26:37)
    at Compiler.<anonymous> (/Users/tsm/Sites/__projects/backbone-starter/node_modules/webpack/lib/Compiler.js:193:12)
    at Compiler.emitRecords (/Users/tsm/Sites/__projects/backbone-starter/node_modules/webpack/lib/Compiler.js:282:37)
    at Compiler.<anonymous> (/Users/tsm/Sites/__projects/backbone-starter/node_modules/webpack/lib/Compiler.js:187:11)
    at /Users/tsm/Sites/__projects/backbone-starter/node_modules/webpack/lib/Compiler.js:275:11
    at Compiler.applyPluginsAsync (/Users/tsm/Sites/__projects/backbone-starter/node_modules/tapable/lib/Tapable.js:60:69)
    at Compiler.afterEmit (/Users/tsm/Sites/__projects/backbone-starter/node_modules/webpack/lib/Compiler.js:272:8)

npm ERR! Darwin 15.2.0
npm ERR! argv "/Users/tsm/.nvm/versions/node/v5.1.0/bin/node" "/Users/tsm/.nvm/versions/node/v5.1.0/bin/npm" "run" "demo"
npm ERR! node v5.1.0
npm ERR! npm  v3.3.12
npm ERR! code ELIFECYCLE
npm ERR! marionette-starter@1.0.0 demo: `gulp production && gulp server`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the marionette-starter@1.0.0 demo script 'gulp production && gulp server'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the marionette-starter package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     gulp production && gulp server
npm ERR! You can get their info via:
npm ERR!     npm owner ls marionette-starter
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/tsm/Sites/__projects/backbone-starter/npm-debug.log
```

npm-debug/log file

``` sh
0 info it worked if it ends with ok
1 verbose cli [ '/Users/tsm/.nvm/versions/node/v5.1.0/bin/node',
1 verbose cli   '/Users/tsm/.nvm/versions/node/v5.1.0/bin/npm',
1 verbose cli   'run',
1 verbose cli   'demo' ]
2 info using npm@3.3.12
3 info using node@v5.1.0
4 verbose run-script [ 'predemo', 'demo', 'postdemo' ]
5 info lifecycle marionette-starter@1.0.0~predemo: marionette-starter@1.0.0
6 silly lifecycle marionette-starter@1.0.0~predemo: no script for predemo, continuing
7 info lifecycle marionette-starter@1.0.0~demo: marionette-starter@1.0.0
8 verbose lifecycle marionette-starter@1.0.0~demo: unsafe-perm in lifecycle true
9 verbose lifecycle marionette-starter@1.0.0~demo: PATH: /Users/tsm/.nvm/versions/node/v5.1.0/lib/node_modules/npm/bin/node-gyp-bin:/Users/tsm/Sites/__projects/backbone-starter/node_modules/.bin:/Users/tsm/.nvm/versions/node/v5.1.0/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/tsm/bin:/Users/tsm/gsutil
10 verbose lifecycle marionette-starter@1.0.0~demo: CWD: /Users/tsm/Sites/__projects/backbone-starter
11 silly lifecycle marionette-starter@1.0.0~demo: Args: [ '-c', 'gulp production && gulp server' ]
12 silly lifecycle marionette-starter@1.0.0~demo: Returned: code: 1  signal: null
13 info lifecycle marionette-starter@1.0.0~demo: Failed to exec demo script
14 verbose stack Error: marionette-starter@1.0.0 demo: `gulp production && gulp server`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (/Users/tsm/.nvm/versions/node/v5.1.0/lib/node_modules/npm/lib/utils/lifecycle.js:232:16)
14 verbose stack     at emitTwo (events.js:87:13)
14 verbose stack     at EventEmitter.emit (events.js:172:7)
14 verbose stack     at ChildProcess.<anonymous> (/Users/tsm/.nvm/versions/node/v5.1.0/lib/node_modules/npm/lib/utils/spawn.js:24:14)
14 verbose stack     at emitTwo (events.js:87:13)
14 verbose stack     at ChildProcess.emit (events.js:172:7)
14 verbose stack     at maybeClose (internal/child_process.js:818:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
15 verbose pkgid marionette-starter@1.0.0
16 verbose cwd /Users/tsm/Sites/__projects/backbone-starter
17 error Darwin 15.2.0
18 error argv "/Users/tsm/.nvm/versions/node/v5.1.0/bin/node" "/Users/tsm/.nvm/versions/node/v5.1.0/bin/npm" "run" "demo"
19 error node v5.1.0
20 error npm  v3.3.12
21 error code ELIFECYCLE
22 error marionette-starter@1.0.0 demo: `gulp production && gulp server`
22 error Exit status 1
23 error Failed at the marionette-starter@1.0.0 demo script 'gulp production && gulp server'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the marionette-starter package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     gulp production && gulp server
23 error You can get their info via:
23 error     npm owner ls marionette-starter
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,4 +1,7 @@
 # Changelog
+## 4.0.2
+- Add `devtool`, `uglifyJsPlugin`, and `definePlugin` environment options
+- [Autoset `uglifyJsPlugin.sourceMap` to `true`](https://github.com/webpack/webpack/issues/2704#issuecomment-228860162) if `production.devtool` is defined
 
 ## 4.0.1
 - add watchOptions to browserSync config [#429](https://github.com/vigetlabs/blendid/pull/429)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,7 +1,8 @@
 # Changelog
-## 4.0.2
+## 4.1.0
 - Add `devtool`, `uglifyJsPlugin`, and `definePlugin` environment options
 - [Autoset `uglifyJsPlugin.sourceMap` to `true`](https://github.com/webpack/webpack/issues/2704#issuecomment-228860162) if `production.devtool` is defined
+- Add `publicPath` to Craft task-config.js [#432](https://github.com/vigetlabs/blendid/issues/432)
 
 ## 4.0.1
 - add watchOptions to browserSync config [#429](https://github.com/vigetlabs/blendid/pull/429)

--- a/README.md
+++ b/README.md
@@ -167,10 +167,30 @@ Define additional webpack plugins that should be used in all environments
 Define additional webpack loaders that should be used in all environments. Adds to `webpackConfig.module.rules`
 
 #### `development`, `production`
-Define additional webpack plugins and loaders for development or production environments
+Specify additional environment specific configuration to be merged in with Blendid's defaults
+
+- [`devtool`](https://webpack.js.org/configuration/devtool/#devtool)
+- [`plugins`](https://webpack.js.org/concepts/plugins/)
+- [`loaders`](https://webpack.js.org/concepts/loaders/)
+
+_Production Only:_
+
+- [`uglifyJsPlugin`](https://webpack.js.org/plugins/uglifyjs-webpack-plugin/#options)
+- [`definePlugin`](https://webpack.js.org/plugins/define-plugin)
+
+Note that if `devtool` is set in production, Blendid will automatically[set to `uglifyJsPlugin.sourceMap` to `true`](https://github.com/webpack/webpack/issues/2704#issuecomment-228860162).
+
+**Example:**
 
 ```js
-development: {
+production: {
+  devtool: 'hidden-source-map',
+  uglifyJsPlugin: {
+    extractComments: true
+  },
+  definePlugin: {
+    SOME_API_KEY: 'abcdefg'
+  },
   plugins: (webpack) => { return [ new webpack.IgnorePlugin(/jsdom$/) ] },
   loaders: [] // Adds to `webpackConfig.module.rules`
 }

--- a/gulpfile.js/lib/task-defaults.js
+++ b/gulpfile.js/lib/task-defaults.js
@@ -11,6 +11,7 @@ module.exports = {
       quiet: true,
       react: false
     },
+    devtool: 'eval-cheap-module-source-map',
     babelLoader: {
       // "test" is derived from TASK_CONFIG.javascripts.extensions
       // "options" is derived from TASK_CONFIG.javascripts.babel
@@ -19,6 +20,16 @@ module.exports = {
     },
     babel: {
       presets: [["es2015", { "modules": false }], 'stage-1']
+    },
+    development: {},
+    production: {
+      devtool: false,
+      uglifyJsPlugin: {},
+      definePlugin: {
+        'process.env': {
+          'NODE_ENV': JSON.stringify('production')
+        }
+      }
     }
   },
 

--- a/gulpfile.js/lib/webpack-multi-config.js
+++ b/gulpfile.js/lib/webpack-multi-config.js
@@ -52,7 +52,7 @@ module.exports = function (env) {
   }
 
   if (env === 'development') {
-    webpackConfig.devtool = TASK_CONFIG.javascripts.devtool || 'eval-cheap-module-source-map'
+    webpackConfig.devtool = TASK_CONFIG.javascripts.development.devtool || TASK_CONFIG.javascripts.devtool || false
     webpackConfig.output.pathinfo = true
 
     // Create new entry object with webpack-hot-middleware and react-hot-loader (if enabled)
@@ -78,13 +78,15 @@ module.exports = function (env) {
       webpackConfig.plugins.push(new webpackManifest(PATH_CONFIG.javascripts.dest, PATH_CONFIG.dest))
     }
 
+    const uglfiyConfig = TASK_CONFIG.javascripts.production.uglifyJsPlugin
+    if(TASK_CONFIG.javascripts.production.devtool) {
+      uglfiyConfig.sourceMap = true
+    }
+
+    webpackConfig.devtool = TASK_CONFIG.javascripts.production.devtool
     webpackConfig.plugins.push(
-      new webpack.DefinePlugin({
-        'process.env': {
-          'NODE_ENV': JSON.stringify('production')
-        }
-      }),
-      new webpack.optimize.UglifyJsPlugin(),
+      new webpack.DefinePlugin(TASK_CONFIG.javascripts.production.definePlugin),
+      new webpack.optimize.UglifyJsPlugin(uglfiyConfig),
       new webpack.NoEmitOnErrorsPlugin()
     )
   }

--- a/gulpfile.js/lib/webpackManifest.js
+++ b/gulpfile.js/lib/webpackManifest.js
@@ -1,18 +1,20 @@
-var path = require('path')
-var fs   = require('fs')
+const path = require('path')
+const fs   = require('fs')
 
 module.exports = function(jsDest, dest, filename) {
   filename = filename || 'rev-manifest.json'
 
   return function() {
-    this.plugin("done", function(stats) {
-      var stats    = stats.toJson()
-      var chunks   = stats.assetsByChunkName
-      var manifest = {}
+    this.plugin("done", function(statsObject) {
+      const stats    = statsObject.toJson()
+      const chunks   = stats.assetsByChunkName
+      const manifest = {}
 
-      for (var key in chunks) {
-        var originalFilename = key + '.js'
-        manifest[path.join(jsDest, originalFilename)] = path.join(jsDest, chunks[key])
+      for (let key in chunks) {
+        const originalFilename = key + '.js'
+        // https://github.com/vigetlabs/blendid/issues/232#issuecomment-171963233
+        const chunkName = typeof chunks[key] === 'string' ? chunks[key] : chunks[key][0]
+        manifest[path.join(jsDest, originalFilename)] = path.join(jsDest, chunkName)
       }
 
       fs.writeFileSync(

--- a/gulpfile.js/tasks/rev/rev-assets.js
+++ b/gulpfile.js/tasks/rev/rev-assets.js
@@ -6,7 +6,7 @@ var revNapkin = require('gulp-rev-napkin');
 // 1) Add md5 hashes to assets referenced by CSS and JS files
 gulp.task('rev-assets', function() {
   // Ignore files that may reference assets. We'll rev them next.
-  var ignoreThese = '!' + path.resolve(process.env.PWD, PATH_CONFIG.dest,'**/*+(css|js|json|html)')
+  var ignoreThese = '!' + path.resolve(process.env.PWD, PATH_CONFIG.dest,'**/*+(css|js|map|json|html)')
 
   return gulp.src([path.resolve(process.env.PWD, PATH_CONFIG.dest,'**/*'), ignoreThese])
     .pipe(rev())

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "blendid",
-  "version": "4.0.2-beta",
+  "version": "4.1.0",
   "description": "(formerly gulp-starter) A full featured configurable gulp asset pipeline and static site builder",
   "license": "MIT",
   "engines": {

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "blendid",
-  "version": "4.0.1",
+  "version": "4.0.2-beta",
   "description": "(formerly gulp-starter) A full featured configurable gulp asset pipeline and static site builder",
   "license": "MIT",
   "engines": {

